In [4]:
import pandas as pd
import numpy as np
from sklearn.model_selection import ShuffleSplit, cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression

### Preuzmite dataset sa linka koji vam je dostupan u datoteci datasets_klasifikacija_binarni_p4.txt. Morat ćete se prijaviti sa Google računom na Kaggle poslužitelj, da bi mogli preuzeti dataset. Proučite dataset i definirajte što je target varijabla. Ako je ne pronalazite, izaberite neku varijablu za koju postoje samo dvije vrijednosti u stupcu. Iz dataseta izbacite sve značajke koje nisu numeričke, a ako negdje nedostaje vrijednost umetnite medijan. Koristeći metodu cross_val_score (za cv koristite ShuffleSplit(), a testna veličina neka bude 20%) usporedite vrijednosti "roc_auc"  algoritama stabla odlučivanja i logističke regresije, pa ispišite dobivene nizove "roc_auc". Radi se o površini ispod krivulje klasifikatora i te vrijednosti bi trebale biti veće od 0,5. Na kraju izračunajte prosjeke površina ispod krivulje za oba modela i u komentar navedite koji je model bolji.

In [5]:
df = pd.read_csv('/content/Car_Insurance_Claim.csv')
df

,ID,AGE,GENDER,RACE,DRIVING_EXPERIENCE,EDUCATION,INCOME,CREDIT_SCORE,VEHICLE_OWNERSHIP,VEHICLE_YEAR,MARRIED,CHILDREN,POSTAL_CODE,ANNUAL_MILEAGE,VEHICLE_TYPE,SPEEDING_VIOLATIONS,DUIS,PAST_ACCIDENTS,OUTCOME
0,569520,65+,female,majority,0-9y,high school,upper class,0.629027,1.0,after 2015,0.0,1.0,10238,12000.0,sedan,0,0,0,0.0
1,750365,16-25,male,majority,0-9y,none,poverty,0.357757,0.0,before 2015,0.0,0.0,10238,16000.0,sedan,0,0,0,1.0
2,199901,16-25,female,majority,0-9y,high school,working class,0.493146,1.0,before 2015,0.0,0.0,10238,11000.0,sedan,0,0,0,0.0
3,478866,16-25,male,majority,0-9y,university,working class,0.206013,1.0,before 2015,0.0,1.0,32765,11000.0,sedan,0,0,0,0.0
4,731664,26-39,male,majority,10-19y,none,working class,0.388366,1.0,before 2015,0.0,0.0,32765,12000.0,sedan,2,0,1,1.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9995,323164,26-39,female,majority,10-19y,university,upper class,0.582787,1.0,before 2015,0.0,0.0,10238,16000.0,sedan,0,0,1,0.0
9996,910346,26-39,female,majority,10-19y,none,middle class,0.522231,1.0,after 2015,0.0,1.0,32765,NaN,sedan,1,0,0,0.0
9997,468409,26-39,male,majority,0-9y,high school,middle class,0.470940,1.0,before 2015,0.0,1.0,10238,14000.0,sedan,0,0,0,0.0
9998,903459,26-39,female,majority,10-19y,high school,poverty,0.364185,0.0,before 2015,0.0,1.0,10238,13000.0,sedan,2,0,1,1.0


In [6]:
# target varijabla
y = df["OUTCOME"]

X = df.drop(columns=[ "ID", "OUTCOME", "AGE", "GENDER", "RACE", "DRIVING_EXPERIENCE", "EDUCATION", "INCOME","VEHICLE_YEAR", "VEHICLE_TYPE" ])

X = X.fillna(X.median()) # znam iz prethodnog zadatka koje kolone treba izbaciti te da imam praznih ćelija


dt = DecisionTreeClassifier(random_state=42)
log_reg = LogisticRegression(max_iter=5000, random_state=42)

cv = ShuffleSplit(n_splits=10, test_size=0.2, random_state=42)

dt_scores = cross_val_score(dt, X, y, cv=cv, scoring='roc_auc')
log_reg_scores = cross_val_score(log_reg, X, y, cv=cv, scoring='roc_auc')

print("Decision Tree AUC scores:", dt_scores.mean())
print("Logistic Regression AUC scores:", log_reg_scores.mean())



Decision Tree AUC scores: 0.6998765370601171
Logistic Regression AUC scores: 0.8428147513048874


## Zaključak: Oba score-a su iznad 0.5, ali je Logistic Regression bolja u ovom slučaju jer joj score iznosi 0.8428 u odnasu na Decision Tree čiji score iznosi 0.6998
